pkg/util/codec/codec.go

```diff
@@ -423,6 +424,175 @@ const (
 	KeepVarColumnLength
 )
 
+// PreAllocForSerializedKeyBuffer estimates key length
+func PreAllocForSerializedKeyBuffer(buildKeyIndex []int, chk *chunk.Chunk, tps []*types.FieldType, usedRows []int, filterVector []bool, nullVector []bool, serializeModes []SerializeMode, serializedKeysVectorBuffer [][]byte, memoryUsagePerRow []int64) ([]int64, error) {
+	rowNum := len(usedRows)
+	if cap(memoryUsagePerRow) < rowNum {
+		memoryUsagePerRow = make([]int64, rowNum)
+	} else {
+		memoryUsagePerRow = memoryUsagePerRow[:rowNum]
+		for i := range rowNum {
+			memoryUsagePerRow[i] = 0
+		}
+	}
+
+	for i, idx := range buildKeyIndex {
+		column := chk.Column(idx)
+		canSkip := func(index int) bool {
+			if column.IsNull(index) {
+				nullVector[index] = true
+			}
+			return (filterVector != nil && !filterVector[index]) || (nullVector != nil && nullVector[index])
+		}
+
+		switch tps[i].GetType() {
+		case mysql.TypeTiny, mysql.TypeShort, mysql.TypeInt24, mysql.TypeLong, mysql.TypeLonglong, mysql.TypeYear:
+			for j := range memoryUsagePerRow {
+				memoryUsagePerRow[j] += size.SizeOfByte + 8
+			}
+		case mysql.TypeFloat, mysql.TypeDouble:
+			for j := range memoryUsagePerRow {
+				memoryUsagePerRow[j] += int64(sizeFloat64)
+			}
+		case mysql.TypeVarchar, mysql.TypeVarString, mysql.TypeString, mysql.TypeBlob, mysql.TypeTinyBlob, mysql.TypeMediumBlob, mysql.TypeLongBlob:
+			collator := collate.GetCollator(tps[i].GetCollate())
+
+			for j, physicalRowIndex := range usedRows {
+				if canSkip(physicalRowIndex) {
+					continue
+				}
+				strLen := int64(len(column.GetBytes(physicalRowIndex)) * collator.MaxLenOneByte())
+				if serializeModes[i] == KeepVarColumnLength {
+					strLen += int64(sizeUint32)
+				}
+
+				memoryUsagePerRow[j] += strLen
+			}
+		case mysql.TypeDate, mysql.TypeDatetime, mysql.TypeTimestamp:
+			for j := range memoryUsagePerRow {
+				memoryUsagePerRow[j] += int64(sizeUint64)
+			}
+		case mysql.TypeDuration:
+			for j := range memoryUsagePerRow {
+				memoryUsagePerRow[j] += 8
+			}
+		case mysql.TypeNewDecimal:
+			ds := column.Decimals()
+			for j, physicalRowindex := range usedRows {
+				if canSkip(physicalRowindex) {
+					continue
+				}
+
+				elemLen := int64(0)
+
+				// Buffer length for all decimal in one column should be same
+				b, err := ds[physicalRowindex].ToHashKey()
```

Comment:
i think you can use `DecimalBinSize` to calculate the length, so there is no need to actually encode the decimal here